bftengine/src/bftengine/messages/ReplicasRestartReadyProofMsg.cpp

```diff
@@ -0,0 +1,136 @@
+                                                           const concordUtils::SpanContext& spanContext)
+    : MessageBase(senderId,
+                  MsgCode::ReplicasRestartReadyProof,
+                  spanContext.data().size(),
+                  ReplicaConfig::instance().getmaxExternalMessageSize() - spanContext.data().size()) {
+  b()->genReplicaId = senderId;
+  b()->seqNum = seqNum;
+  b()->elementsCount = 0;
+  b()->locationAfterLast = 0;
+  b()->sigLength = sigLen;
+  std::memcpy(body() + sizeof(Header), spanContext.data().data(), spanContext.data().size());
+}
+const uint32_t ReplicasRestartReadyProofMsg::getBodySize() const {
+  uint32_t bodySize = b()->locationAfterLast;
+  if (bodySize == 0) bodySize = sizeof(Header) + spanContextSize();
+  return bodySize;
+}
+
+ReplicasRestartReadyProofMsg* ReplicasRestartReadyProofMsg::create(ReplicaId id,
+                                                                   SeqNum s,
+                                                                   const concordUtils::SpanContext& spanContext) {
+  auto sigManager = SigManager::instance();
+  const size_t sigLen = sigManager->getMySigLength();
+  ReplicasRestartReadyProofMsg* m = new ReplicasRestartReadyProofMsg(id, s, sigLen, spanContext);
+  return m;
+}
+
+void ReplicasRestartReadyProofMsg::addElement(std::unique_ptr<ReplicaRestartReadyMsg>& restartMsg) {
+  if (b()->locationAfterLast == 0)  // if this is the first element
+  {
+    ConcordAssert(b()->elementsCount == 0);
+    b()->locationAfterLast = sizeof(Header) + spanContextSize();
+  }
+  uint32_t requiredSpace = b()->locationAfterLast + restartMsg->size();
+  ConcordAssertLE((size_t)(requiredSpace + SigManager::instance()->getMySigLength()), (size_t)internalStorageSize());
+  std::memcpy(body() + b()->locationAfterLast, restartMsg->body(), restartMsg->size());
+  b()->elementsCount += 1;
+  b()->locationAfterLast = requiredSpace;
+}
+// +-----------------------------------------------------------------------+--------------+
+// | Msg header(genReplicaId, seqNum, eleCount, locationAfterLast, sigLen) | Span Context |
+// +-----------------------------------------------------------------------+--------------+
+// |  Element1(replicaId, seqNum, sigLen, sigBody)  |
+// +------------------------------------------------+
+// |     .................................          |
+// +------------------------------------------------+
+// |  Elementn(replicaId, seqNum, sigLen, sigBody)  |
+// +------------------------------------------------+
+// |  Signature of the previous fields              |
+// +------------------------------------------------+
+
+void ReplicasRestartReadyProofMsg::finalizeMessage() {
+  auto bodySize = getBodySize();
+  auto sigManager = SigManager::instance();
+  auto sigSize = sigManager->getMySigLength();
+  setMsgSize(bodySize + sigSize);
+  shrinkToFit();
+  sigManager->sign(body(), bodySize, body() + bodySize, sigSize);
```

Comment:
We don't need to sign the proof message (because it contains signed "readyToRestart" messages